src/jobs.rs

```diff
@@ -486,6 +505,156 @@ impl CommitJobInstance {
+                    return Ok(Some((storefront_info, build_ref.commit.clone())));
+                }
+
+                /* gzip encode the new appstream file */
+                let mut s = vec![];
+                GzEncoder::new(&mut s, Compression::default())
+                    .write_all(new_appstream.as_bytes())?;
+
+                /* Make sure all the file attributes are correct */
+                fileinfo.set_size(s.len().try_into().expect("integer overflow on file size"));
+                fileinfo.set_name(&format!("{}.xml.gz", app_id));
+                fileinfo.set_file_type(FileType::Regular);
+                fileinfo.set_attribute_uint32("unix::uid", 0);
+                fileinfo.set_attribute_uint32("unix::gid", 0);
+                fileinfo.set_attribute_uint32("unix::mode", 0o100644);
+
+                repo.prepare_transaction::<Cancellable>(None)?;
+
+                /* Write the new appstream file to the repo */
+                let checksum =
+                    repo.write_regfile_inline::<Cancellable>(None, 0, 0, 0o100644, None, &s, None)?;
+
+                /* Edit the MutableTree with a reference to the new appstream file */
+                mtree_lookup(&mtree, &["files", "share", "app-info", "xmls"])?
+                    .1
+                    .ok_or("file not found")?
+                    .replace_file(&format!("{}.xml.gz", app_id), &checksum)?;
+
+                let repo_file = repo.write_mtree::<Cancellable>(&mtree, None)?;
+
+                /* Copy the original commit metadata. Leave out extended attributes, that's just the signature, which
+                won't be valid when we rewrite the commit (and we will sign the resulting commit ourselves anyway) */
+                let (commit_metadata, _state) = repo.load_commit(&build_ref.commit)?;
+                let metadata = commit_metadata.child_get::<VariantDict>(0);
+                let subject = &commit_metadata.child_get::<String>(3);
+                let body = &commit_metadata.child_get::<String>(4);
+                let time = libostree::commit_get_timestamp(&commit_metadata);
+
+                /* Write a new commit with the new dirtree but (mostly) the same metadata */
+                let commit = repo.write_commit_with_time::<Cancellable>(
+                    Some(&build_ref.commit),
+                    Some(subject),
+                    Some(body),
+                    Some(&metadata.end()),
+                    repo_file.dynamic_cast_ref().unwrap(),
+                    time,
+                    None,
+                )?;
+
+                repo.commit_transaction::<Cancellable>(None)?;
+
+                job_log_and_info(
+                    self.job_id,
+                    conn,
+                    &format!("Rewriting commit {} -> {}", build_ref.commit, commit),
+                );
+
+                Ok(Some((storefront_info, commit.to_string())))
+            } else {
+                Ok(None)
```

Comment:
This function is long; perhaps use e.g. rust-analyzer to extract a function for the code in the `if`?  Then you could do just e.g.:

```
config.storefront_info_endpoint.as_ref().map(rewrite_commit).transpose()?
```